organization/src/main/scala/com/improving/app/organization/domain/Organization.scala

```diff
@@ -1,6 +1,206 @@
+    event match {
+      case OrganizationEvent.Empty => state
+      case event: OrganizationEstablished =>
+        state match {
+          case UninitializedState => DraftState(info = event.organizationInfo.get, metaInfo = event.metaInfo.get)
+          case _: DraftState => state
+          case _: ActiveState => state
+          case _: SuspendedState => state
+        }
+      case event: OrganizationActivated =>
+        state match {
+          case x: DraftState => ActiveState(x.info, event.metaInfo.get)
+          case _: ActiveState => state
+          case x: SuspendedState => ActiveState(x.info, event.metaInfo.get)
+          case UninitializedState => UninitializedState
+        }
+      case event: OrganizationSuspended =>
+        state match {
+          case x: DraftState => SuspendedState(x.info, event.metaInfo.get)
+          case x: ActiveState => SuspendedState(x.info, event.metaInfo.get)
+          case x: SuspendedState => SuspendedState(x.info, event.metaInfo.get)
+          case UninitializedState => UninitializedState
+        }
+      case _: OrganizationTerminated =>
+        ???
+    }
+  }
+
+  private def updateMetaInfo(metaInfo: OrganizationMetaInfo, lastUpdatedByOpt: Option[MemberId]): OrganizationMetaInfo = {
+    metaInfo.copy(lastUpdatedBy = lastUpdatedByOpt, lastUpdated = Some(Timestamp(Instant.now())))
+  }
+
+  private def establishOrganization(establishOrganization: EstablishOrganization): Either[Error, OrganizationEvent] = {
+    val maybeValidationError = applyAllValidators[EstablishOrganization](Seq(
+      c => required("organization id", organizationIdValidator)(c.organizationId),
+      c => required("on behalf of", memberIdValidator)(c.onBehalfOf)
+    ))(establishOrganization)
+    if(maybeValidationError.isDefined) {
+      Left(maybeValidationError.get)
+    } else {
+      val maybeOrganizationInfoError = required("organization info", completeOrganizationInfoValidator)(establishOrganization.organizationInfo)
+      if (maybeOrganizationInfoError.isDefined) {
+        Left(maybeOrganizationInfoError.get)
+      } else {
+        val organizationInfo = establishOrganization.organizationInfo.get
+
+        val newMetaInfo = OrganizationMetaInfo(
+          createdOn = Some(Timestamp(Instant.now())),
+          createdBy = Some(establishOrganization.getOnBehalfOf)
+        )
+
+        Right(OrganizationEstablished(
+          organizationId = establishOrganization.organizationId,
+          metaInfo = Some(newMetaInfo),
+          organizationInfo = Some(organizationInfo)
+        ))
+      }
+    }
+  }
+
+  private def activateOrganization(
+                                    state: DraftState,
+                                    activateOrganization: ActivateOrganization,
+                                  ): Either[Error, OrganizationEvent] = {
+    val maybeValidationError = applyAllValidators[ActivateOrganization](Seq(
```

Comment:
Also just commenting here for just some thoughts on your next PR of implementing more functions in Organizations.

I've noticed that you're validating organization id and "on behalf of" again here when you've already done it in the service impl. Shouldn't there be only one? Where is the proper place to put the validation? Anyway, this is just a comment. Feel free to resolve after reading this.